package dsbproto.execution;

import "model.proto";


// Message type identifiers.
enum MessageType
{
    // Handshake
    MSG_DENIED       =  0;
    MSG_HELLO        =  1;

    // Commands
    MSG_SETUP        = 10;
    MSG_SET_VARS     = 11;
    MSG_STEP         = 12;
    MSG_ACCEPT_STEP  = 13;
    MSG_TERMINATE    = 14;
    MSG_DESCRIBE     = 15;
    MSG_SET_PEERS    = 16;

    // Responses
    MSG_ERROR        = 30;
    MSG_READY        = 31;
    MSG_STEP_OK      = 32;
    MSG_STEP_FAILED  = 33;
}

// The body of an ERROR message.
message ErrorInfo
{
    enum Code
    {
        UNKNOWN_ERROR   =  0;
        INVALID_REQUEST =  1;
    }

    optional Code code = 1 [default = UNKNOWN_ERROR];
    optional string details = 2;
}

// Information sent by a slave about itself
message SlaveDescription
{
    required model.SlaveTypeDescription type_description = 1;
}

// The ID number and a value for one of a slave's variables.
message SlaveVariableSetting
{
    required uint32 variable_id = 1;
    optional model.ScalarValue value = 2;
    optional model.Variable connected_output = 3;
}

// The body of a SETUP message
message SetupData
{
    required uint32 slave_id = 1;
    required double start_time = 2;
    optional double stop_time = 3;
    optional string execution_name = 4;
    optional string slave_name = 5;
}

// A message that is sent by the master to a slave to set some of its variables.
message SetVarsData
{
    repeated SlaveVariableSetting variable = 1;
}

// The body of a STEP message
message StepData
{
    required int32 step_id = 1;
    required double timepoint = 2;
    required double stepsize = 3;
}

// The body of a SET_PEERS message
message SetPeersData
{
    repeated string peer = 1;
}
